storage/badger.go

```diff
@@ -0,0 +1,186 @@
+package storage
+
+import (
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pkg/errors"
+	"os"
+)
+
+const (
+	// PerformanceMemTableSize is 3072 MB. Increases the maximum
+	// amount of data we can commit in a single transaction.
+	PerformanceMemTableSize = 3072 << 20
+
+	// PerformanceLogValueSize is 256 MB.
+	PerformanceLogValueSize = 256 << 20
+)
+
+type BadgerDatabase struct {
+	db            *badger.DB
+	opts          badger.Options
+	useWriteBatch bool
+}
+
+func NewBadgerDatabase(opts badger.Options, useWriteBatch bool) *BadgerDatabase {
+	return &BadgerDatabase{
+		db:            nil,
+		opts:          opts,
+		useWriteBatch: useWriteBatch,
+	}
+}
+
+func (bdb *BadgerDatabase) Setup() error {
+	db, err := badger.Open(bdb.opts)
+	if err != nil {
+		return errors.Wrapf(err, "Setup:")
+	}
+	bdb.db = db
+	return nil
+}
+
+func (bdb *BadgerDatabase) Update(fn func(Transaction) error) error {
+	var wb *badger.WriteBatch
+	if bdb.useWriteBatch {
+		wb = bdb.db.NewWriteBatch()
+	}
+
+	err := bdb.db.Update(func(txn *badger.Txn) error {
+		T := NewBadgerTransaction(txn, wb)
+		return fn(T)
+	})
+	if err != nil {
+		return errors.Wrapf(err, "Update:")
+	}
+
+	if bdb.useWriteBatch {
+		err = wb.Flush()
+		if err != nil {
+			return errors.Wrapf(err, "Update: Problem flushing write batch")
+		}
+		wb.Cancel()
+	}
+	return nil
```

Comment:
Just a nit. It's worth getting into the habit of un-nesting if blocks where possible.
```suggestion
	if !bdb.useWriteBatch {
		return nil
	}
	
	if err = wb.Flush(); err != nil {
		return errors.Wrapf(err, "Update: Problem flushing write batch")
	}
	
	wb.Cancel()
```

Reply:
agreed^